Pull request for issue #13

cyrilleponcelet opened the pull request:
Take into account several position and attitude sensors connected and add a few missing variables

cyrilleponcelet committed:
Take into account several position and attitude sensors connected and add a few missing variables

gavinmacaulay commented:
A few comments:
- Is it useful to specify a preferred position and attitude group for when there are more than one? This would help programs that don't know which sensors to use, for generic use.
- Should it be more clearly stated that the name of the group _must_ be the same as listed in MRU_ids and position_ids?
- I wonder if the offset_{x,y,} and rotation_{x,y,z} variables are more appropriately put into the subgroup that contains the data from that sensor?
- Are there sensors other than position and attitude that we should explicitly mention? Or at least mention that additional types of sensors can be included using a similar structure?

cyrilleponcelet commented:
- Very good idea to set a preferred position. Most of the system only have one position and one attitude sensor so choice is obvious. As far as I know for bathymetry change of sensor input is like a change of configuration (it creates a new file). So having the preferred position definitely make sense, since for most system it is not dynamic. Do you want to put this information in the sensor group ? I would rather put a variable in the sonar beamGroup, this variable will be time indexed if the system change over time or will have a global value if it does not change with time. 

- Point 2 agreed.

- I liked the idea to have the platform description really separated from the sensor data. I would keep it like this (or you will need to mode sonar description to the sonar group too)

- Yes we could mention that additional types of sensors can be included using a similar structure. May be we could add a list for naming the groups. On our systems, I know as external sensors : Clock, Depth (vertical depth), Surface Sound Speed (but maybe it should be in environnement group). Is that ok for you 

If all is good for you I will amend the pull request

gavinmacaulay commented:
Sounds good. Please amend the pull request.

- Agreed - the preferred sensor information seems best in the beamGroup.
- :)
- Sounds good to keep separation between platform description and sensor data
- A suggested list of names might help. But clarify in the text that this group is for sensors that measure something about the platform (so sound speed belongs in Environment).

cyrilleponcelet commented:
Great, I'll do that very soon.
For the last point, just to be clear. If I want to record a sensor giving "Vertical Depth", is it in Environment group ? I think so

gavinmacaulay commented:
Yes, depth from a separate sensor is part of the environment, not the platform (although depth could be derived from data in a beam group and that would make sense to be under the relevant beam group, aka ADCP and multibeam proposal).

cyrilleponcelet committed:
Add preferred sensor to beamGroup, try to make clear the Platform group description, clean up attitude subgroup from crs notion

cyrilleponcelet commented:
I have take into account all comments and tried to make the platform description more clear.
I also made some clean up in the attitude subgroup (removed a crs notion that I will submit later)

gavinmacaulay commented on SONAR-netCDF4/docs/tableBeamGroup1.adoc:
```diff
@@ -143,4 +145,12 @@
+|int active_position_sensor(ping_time)|MA |Indicate the index of the position sensor used at the time of the ping to compute the platform position.
+|:valid_min = "0" | |
+|:long_name = "Active position sensor index"| |
+ =========================================================================================================================================================================================================================================================================================================================================================
```
The end of table marker needs a | before the ='s (generates an error when running through asciidoctor otherwise).

cyrilleponcelet replied:
My mistake, sorry

cyrilleponcelet committed:
Add missing | at the end of table

cyrilleponcelet committed:
Update tablePlatform.adoc

rename suggested subgroups for position and attitude sensor data. It is not possible to have the same name for a group as a dimension name. Names were differing only with a capital letter, which can be source of confusion. Use plural and capital letter instead

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
-
-Optionally, /Platform subgroups can be included to store data from individual instruments. This would provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. The data format is not prescribed, but could, for example, be NMEA-style text and/or numeric values. Subgroups must have one attribute called “description” that provi ...
+This group contains information about the sonar platform (e.g. ship). The netCDF4 group name is */Platform* and is described in <<platformGroupTable>>. 
+Optionally, /Platform subgroups can be included to store data from individual instruments measuring values that are relative to the platform. If those values are relative to the environment around the platform, they must be stored in the Environment group. 
```
```suggestion
Optionally, subgroups of /Platform can be used to store data from individual instruments that provide measurements about the platform (note that if those values are relative to the environment around the platform, they should be stored in the Environment group). 
```

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
-Optionally, /Platform subgroups can be included to store data from individual instruments. This would provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. The data format is not prescribed, but could, for example, be NMEA-style text and/or numeric values. Subgroups must have one attribute called “description” that provi ...
+This group contains information about the sonar platform (e.g. ship). The netCDF4 group name is */Platform* and is described in <<platformGroupTable>>. 
+Optionally, /Platform subgroups can be included to store data from individual instruments measuring values that are relative to the platform. If those values are relative to the environment around the platform, they must be stored in the Environment group. 
+Each platform can be made up with several instruments belonging to the same category. (For example a platform can associated to several position sensors). Each category is defined as a subgroup of the Platform group and data for each sensor are stored in a subgroup of the sensor category. This structure must be used to store any type of sensor. 
```
```suggestion

Each platform can have several instruments belonging to the same category (for example, a platform can have several position sensors). Each category is defined as a subgroup of the Platform group and data for each sensor are stored in a subgroup of the sensor subgroup. This structure can be used to store any type of platform sensor. 
```

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
+This group contains information about the sonar platform (e.g. ship). The netCDF4 group name is */Platform* and is described in <<platformGroupTable>>. 
+Optionally, /Platform subgroups can be included to store data from individual instruments measuring values that are relative to the platform. If those values are relative to the environment around the platform, they must be stored in the Environment group. 
+Each platform can be made up with several instruments belonging to the same category. (For example a platform can associated to several position sensors). Each category is defined as a subgroup of the Platform group and data for each sensor are stored in a subgroup of the sensor category. This structure must be used to store any type of sensor. 
+More specifically individual position and (resp.) attitude instruments are defined individually as subgroups of the */Platform/Position* and (resp.) */Platform/Attitude* subgroups. The name of those subgroups must match the names of their serial number or identifier as defined in <<platformGroupTable>>.
```
```suggestion
For example, data from individual position and attitude instruments are stored as subgroups of the */Platform/Position* and */Platform/Attitude* subgroups. The names of those subgroups must match the names of the instrument's serial number or identifier as defined in <<platformGroupTable>>.
```

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
+Each platform can be made up with several instruments belonging to the same category. (For example a platform can associated to several position sensors). Each category is defined as a subgroup of the Platform group and data for each sensor are stored in a subgroup of the sensor category. This structure must be used to store any type of sensor. 
+More specifically individual position and (resp.) attitude instruments are defined individually as subgroups of the */Platform/Position* and (resp.) */Platform/Attitude* subgroups. The name of those subgroups must match the names of their serial number or identifier as defined in <<platformGroupTable>>.
+The coordinate system convention, including the definition of pitch and roll, is detailed in <<_coordinate_systems>>.
+Each sensor instance group would also provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. 
```
```suggestion

Each sensor instance group would also provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. 
```

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
+Each sensor instance group would also provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. 
+The data format is not prescribed, but could be, for example, be NMEA-style text and/or numeric values. Subgroups must have one attribute called “description” that provides a short description of the contents. Other attributes can be added as desired. The variables under the subgroup should have appropriate names and an attribute that gives the units, where appropriate.
+The following table give the suggested convention recommended to name subgroups for different known kind of sensors or datagrams
+[cols=",,",options="header",]
```
```suggestion
.Suggested subgroup names for platform instruments.
[cols=",,",options="header",]
[[platformInstrumentsTable]]
```

gavinmacaulay commented on SONAR-netCDF4/docs/crr341.adoc:
```diff
@@ -194,15 +194,39 @@ include::tableEnvironment.adoc[]
+The coordinate system convention, including the definition of pitch and roll, is detailed in <<_coordinate_systems>>.
+Each sensor instance group would also provide a standard place to store unprocessed and unparsed instrument data from multiple GPS sensors, motion reference units, etc. 
+The data format is not prescribed, but could be, for example, be NMEA-style text and/or numeric values. Subgroups must have one attribute called “description” that provides a short description of the contents. Other attributes can be added as desired. The variables under the subgroup should have appropriate names and an attribute that gives the units, where appropriate.
+The following table give the suggested convention recommended to name subgroups for different known kind of sensors or datagrams
```
```suggestion
<<platformInstrumentsTable>> gives recommended platform sensor subgroup names.
```

gavinmacaulay commented on SONAR-netCDF4/docs/tablePlatform.adoc:
```diff
@@ -9,110 +8,54 @@
-|:standard_name =  "platform_pitch_angle" | |
-|:units = "arc_degree" | |
-|float :valid_range = –90.0, 90.0 | |
+|string MRU_ids(MRU) |R |MRU serial number or identification name
```
```suggestion
|string MRU_ids(MRU) |R |MRU serial number or identification name. Must be a valid netCDF4 group name.
```

gavinmacaulay commented on SONAR-netCDF4/docs/tablePlatform.adoc:
```diff
@@ -9,110 +8,54 @@
+|string MRU_ids(MRU) |R |MRU serial number or identification name
 | | |
-|float position_offset_x |R |Distance from the platform coordinate system origin to the latitude/longitude position origin along the _x_-axis.
+|string position_ids(position) |R |Position sensor serial number or identification name
```
```suggestion
|string position_ids(position) |R |Position sensor serial number or identification name. Must be a valid netCDF4 group name.
```

gavinmacaulay commented on SONAR-netCDF4/docs/tablePosition_sub_group.adoc:
```diff
@@ -0,0 +1,62 @@
+|:long_name = "longitude" | |
+|double:_FillValue = Double.NaN | |
+| | |
+|float heading(time) |MA |Heading refers to the direction a platform is pointing. This may or may not be the direction that the platform actually travels, which is known as its course or track. Any difference between course and heading is due to the motion of the underlying medium, the air or water, or other effects like skidding or slipping. Heading is typically based on compass directions, so 0 ...
```
Heading is also in the attitude group. How do reading programs decided which heading to use? 

The description field is different to the heading description in attitude. They should be the same (perhaps merge the best parts from both fields?).

cyrilleponcelet replied:
Another thing crossed my mind. What to do with sensors that are both attitude and position sensor ? Any idea ?
Same question arise since attitude sensors can have heading determination too, same as GPS. How to determine what data is used the systems. I think it is system dependent and do not know how to have a generic description for their use.

gavinmacaulay replied:
This comes down to what is an instrument and what is a sensor... Or that instruments can be made up of several sensors... 

We have two aims that conflict a little:
1. store data from multiple sensors (some of which produce the same or similar measurements)
2. provide clear directions on which measurements to use for routine processing of data

V1.0 of the convention did 2, but not 1. The proposed v1.1 currently does 1 but not 2. Perhaps we can have 'recommended_X' variables in each beam_group that gives the path and name of the recommended/suggested position, speed, heading, pitch, roll, etc, variables?

Then, if an instrument provided attitude and position, reading software just needs to look at the 'recommended_X' variables to know where to find things and it doesn't matter where the data are, or what exact type of instrument collected the data.

cyrilleponcelet replied:
Agreed with the distinction between sensor and attitude.
Concerning the capacity to read variables, what's interesting softwares is most of the time to have the attitude and position at the time of the ping. Data in sensor are not stored at the time of the ping but usually at a higher frequency. 
What I will suggest is to have other variables stored in the beamGroup containing position and attitude at the time of the ping (interpolation are usually done by the sounders). (Remind that this is constraint by the fact that the time dimension for the ping is in the beamGroup)
Those variables are the ones used by the sounder and there is no longuer any question about where to find, except for very special case where you need to retrieve source data at higher frequency. 
As soon as you validate this pull request I will create a new one to add those variables. (I could have done that in the same, but I makes it harder to review)
For your information, we use this approach for bathymetry data and we suggested this format for ADCP too.

gavinmacaulay commented on SONAR-netCDF4/docs/tablePosition_sub_group.adoc:
```diff
@@ -0,0 +1,62 @@
+|:units = "degree" | |
+|:long_name = "Platform heading(true)" | |
+| | |
+|float course_over_ground(time) |O |a platform course is the cardinal direction along which the platform or aircraft is to be steered
```
Are we planning to use this data format for aircraft also? ;)

cyrilleponcelet replied:
Oups, my mistake, probably a copy paste (And in my mind a ship is always a kind of aircraft stuck on the surface of water).
I removed this

gavinmacaulay commented on SONAR-netCDF4/docs/tablePosition_sub_group.adoc:
```diff
@@ -0,0 +1,62 @@
+|:units = "m/s" | |
+|float :valid_min = 0.0 | |
+| | |
+|float height_above_reference(time) |MA |height of the platform reference point above reference coordinate system (could be geoid, ellipsoid, or chart datum specified in vertical reference attribute)
```
Is the reference coordinate system defined anywhere?

cyrilleponcelet replied:
The coordinate system is supposed to defined in the vertical_reference attribute. I will comment on this attribute

gavinmacaulay commented on SONAR-netCDF4/docs/tablePosition_sub_group.adoc:
```diff
@@ -0,0 +1,62 @@
+|:units = "m" | |
+|:long_name = "height_above_reference" | |
+| | |
+|int vertical_reference(time) |MA |see 5.6. Horizontal Coordinate Reference Systems, Grid Mappings, and Projections in cf-convention 
```
The description seems to be copied from somewhere, and the reference to 5.6 is then not useful.

cyrilleponcelet replied:
This was a reference to the CF convention, explaining the system reference used. More specifically the lat/lon are pretty clear everything is refered to WGS84 but the height_above_reference need to define its reference coordinate system. I planned to add a variable for this, that's why we have height_above_reference:grid_mapping=vertical_reference attribute.

Usually, we could find height relative to ellipsoid or eight relative to geoid (mean sea level). The purpose of this vertical_reference variable was to define which coordinate system is height referred to. Thanks to your remark I just realized that there is no explicit convention for vertical coordinates. 
I can suggest to change height_above_reference to height_above_ellipsoid and add a variable height_above_geoid. I think at least most system will deliver at least elevation relative to ellipsoid (In NMEA protocol it is defined relative to MSL (geoid) with field to refer back to ellipsoid). But we can let the sensor use one or the other (or both).
What do you think of this proposal ?

gavinmacaulay replied:
I'm a bit out of my depth with this stuff. As long as it follows some conventions, is easy for non-experts to understand, and there is a diagram/figure that explains it, I'm happy :)

cyrilleponcelet replied:
I think height_above_ellipsoid and height_above_geoid is explicit, easy to understand and follow convention so I will create those to variables

cyrilleponcelet replied:
Just pushed my changes. I created to variables : altitude and height_above_reference_ellipsoid to be close to the CF standards. (http://cfconventions.org/Data/cf-standard-names/65/build/cf-standard-name-table.html)

gavinmacaulay commented on SONAR-netCDF4/docs/tableAttitude_sub_group.adoc:
```diff
@@ -0,0 +1,54 @@
+|:standard_name = "time" | |
+|:units = "nanoseconds since 1601-01-01 00:00:00Z" | |
+|Variables | |
+|float distance(time) |O |Distance travelled by the platform from an arbitrary location.
```
Is this part of the attitude instrument? It typically comes from a doppler log or the GPS.

cyrilleponcelet replied:
I moved it to the position group. If it comes from another sensor, it might be added as a new subgroup

gavinmacaulay commented on SONAR-netCDF4/docs/tableAttitude_sub_group.adoc:
```diff
@@ -0,0 +1,54 @@
+|:units = "m" | |
+|:valid_min = "0.0" | |
+| | |
+|float heading(time) |MA |Platform heading. Measured clockwise from north.
```
Does heading belong in the attitude subgroup, or would it be better in the position subgroup?

gavinmacaulay commented:
I've gone through the commits and made some comments and suggestions. Some other items that weren't part of the changes (so I couldn't comment directly on them) or were more general:
- There are MRU_ids, position_ids, and transducer_id. To be consistent we should use all plurals or no plurals. I have no strong opinion on which one we should use.
- I worry a little that the basic information needed to use the acoustic data is now split across two subgroups.
- Should the MRU_id(s), position_id(s), and transducer_id(s) be MA rather than R, since some of the variables in the subgroups are M?
- Supporting multiple instruments of the same type is a big change for reading programs, so I think we should increment the version of the convention before we forget... (to v1.1 I suppose).

gavinmacaulay commented on SONAR-netCDF4/docs/tablePosition_sub_group.adoc:
```diff
@@ -0,0 +1,62 @@
+|:long_name = "height_above_reference" | |
+| | |
+|int vertical_reference(time) |MA |see 5.6. Horizontal Coordinate Reference Systems, Grid Mappings, and Projections in cf-convention 
+|:grid_mapping_name = " " | |
```
This needs a description.

cyrilleponcelet replied:
Same as 2 comments before, commented in the first one

cyrilleponcelet committed:
Update SONAR-netCDF4/docs/crr341.adoc

Co-Authored-By: Gavin Macaulay <gavin@macaulay.co.nz>

cyrilleponcelet committed:
Update SONAR-netCDF4/docs/crr341.adoc

Co-Authored-By: Gavin Macaulay <gavin@macaulay.co.nz>

cyrilleponcelet committed:
Update SONAR-netCDF4/docs/crr341.adoc

Co-Authored-By: Gavin Macaulay <gavin@macaulay.co.nz>

cyrilleponcelet committed:
Update SONAR-netCDF4/docs/tablePlatform.adoc

Co-Authored-By: Gavin Macaulay <gavin@macaulay.co.nz>

cyrilleponcelet committed:
Update SONAR-netCDF4/docs/tablePlatform.adoc

Co-Authored-By: Gavin Macaulay <gavin@macaulay.co.nz>

cyrilleponcelet committed:
Remove aircraf mention from course_over_ground

cyrilleponcelet committed:
Move distance variable from attitude to position subgroup

cyrilleponcelet committed:
Use plural for transducer_ids

cyrilleponcelet committed:
Set transducer_ids,MRU_ids,position_ids to MA

cyrilleponcelet committed:
Update document version to 1.1

cyrilleponcelet committed:
Update Conventions attribute to SONAR-netCDF4-1.1

cyrilleponcelet commented:
>     * There are MRU_ids, position_ids, and transducer_id. To be consistent we should use all plurals or no plurals. I have no strong opinion on which one we should use.

OK, I will use plural. Are you ok with the fact that we sometimes use the word 'MRU' and sometime attitude ?

 
>     * I worry a little that the basic information needed to use the acoustic data is now split across two subgroups.
I do not, the main reason is that next pull request will probably about adding some variables in beamGroup matching position and attitude interpolated at the time of the ping (often interpolation is done by sounder). Most of the time people will use those variables and not the ones in Platform

> 
>     * Should the MRU_id(s), position_id(s), and transducer_id(s) be MA rather than R, since some of the variables in the subgroups are M?
Yes, I do that. (Note that In my opinion every offset and rotation information should be mandatory)

> 
>     * Supporting multiple instruments of the same type is a big change for reading programs, so I think we should increment the version of the convention before we forget... (to v1.1 I suppose).

OK, once we will have Bathymetry, ADCP and other subgroups maybe a v2.0 might be good

gavinmacaulay commented:
We could change MRU to attitude in the variable names, etc, but it would cause hassles for software that supports v1.0 of the convention, so I suggest to leave this slight discontinuity...

cyrilleponcelet commented:
OK to keep MRU names like it is, less work for me

cyrilleponcelet committed:
Refactor heigh_above_reference fields to be close to cf conventions

cyrilleponcelet committed:
Refactor heigh_above_reference fields to be close to cf conventions
